File: ios/Namu/Native/DescriptorVerifier.swift
import CryptoKit
import Foundation

/// Verified, typed view of a signed release descriptor payload (SIG-002).
struct ReleaseDescriptor: Equatable {
  let schema: Int64
  let sequence: Int64
  let modelId: String
  let artifactVersion: String
  let path: String
  let bytes: Int64
  let sha256: String
  let upstreamRepo: String
  let upstreamRevision: String
  let upstreamFilename: String
  let architecture: String
  let quantization: String
  let runtimeBuildIds: [String]
  let minAppBuild: Int64
  let maxAppBuild: Int64
  let promptVersion: String
  let licenseNoticeId: String
  let issuedAtMs: Int64
  let expiresAtMs: Int64
}

struct VerificationKey: Equatable {
  let keyId: String
  let publicKeyB64: String
}

enum DescriptorSource: String {
  case bundled
  case update
}

struct DescriptorProfile: Equatable {
  let architectures: [String]
  let quantizations: [String]

  /// Release builds accept only the locked production artifact family.
  static let production = DescriptorProfile(architectures: ["cohere2"], quantizations: ["Q4_K_M"])
  /// Internal builds also accept the small development fixture (REL-001).
  static let internalFixture = DescriptorProfile(
    architectures: ["cohere2", "llama"], quantizations: ["Q4_K_M", "Q8_0", "F16", "F32"])
}

struct DescriptorContext {
  var keys: [VerificationKey]
  var source: DescriptorSource
  var appBuild: Int64
  var runtimeBuildId: String
  var promptVersions: [String] = NamuConstants.understoodPromptVersions
  var licenseNoticeIds: [String] = NamuConstants.bundledLicenseNoticeIds
  var nowMs: Int64
  var highestSequence: Int64 = 0
  var highestSequencePayloadSha256: String?
  var knownBad: Set<String> = []
  var profile: DescriptorProfile = .production
}

enum DescriptorVerification: Equatable {
  case accepted(ReleaseDescriptor, payloadSha256: String)
  case rejected(code: String, reason: String)
}

/// Mirrors model-release/descriptor/verify.mjs step by step (contract §4).
/// Both must agree on every vector in descriptor-vectors.json.
enum DescriptorVerifier {
  static let maxEnvelopeBytes = 65_536
  static let maxPayloadBytes = 32_768
  static let maxValidityMs: Int64 = 180 * 24 * 3600 * 1000
  static let clockSkewMs: Int64 = 24 * 3600 * 1000

  static func verify(envelope raw: Data, context ctx: DescriptorContext) -> DescriptorVerification {
    func reject(_ code: String, _ reason: String) -> DescriptorVerification {
      .rejected(code: code, reason: reason)
    }
    let invalid = NamuError.signatureInvalid

    // 1. Size limits and strict parse of the envelope.
    guard !raw.isEmpty, raw.count <= maxEnvelopeBytes else { return reject(invalid, "envelope size") }
    guard let envelope = try? StrictJSON.parse(raw), case .object = envelope else {
      return reject(invalid, "envelope parse")
    }
    guard let keyId = envelope.string("key_id"), !keyId.isEmpty,
          let payload = decodeBase64(envelope.string("payload_b64")),
          let signature = decodeBase64(envelope.string("signature_b64")) else {
      return reject(invalid, "envelope fields")
    }
    guard !payload.isEmpty, payload.count <= maxPayloadBytes else { return reject(invalid, "payload size") }
    guard signature.count == 64 else { return reject(invalid, "signature size") }

    // 2. key_id selects a bundled key.
    guard let key = ctx.keys.first(where: { $0.keyId == keyId }) else { return reject(invalid, "unknown key") }
    guard let publicRaw = decodeBase64(key.publicKeyB64), publicRaw.count == 32,
          let publicKey = try? Curve25519.Signing.PublicKey(rawRepresentation: publicRaw) else {
      return reject(invalid, "bundled key")
    }

    // 3. Ed25519 over the exact decoded payload bytes (SIG-002).
    guard publicKey.isValidSignature(signature, for: payload) else { return reject(invalid, "signature") }

    // 4. Only verified bytes are parsed from here on.
    guard let p = try? StrictJSON.parse(payload), case .object = p else {
      return reject(invalid, "payload parse")
    }

    // 5. Field validation. Unknown fields are ignored (SIG-003).
    guard let schema = p.integer("schema"),
          let sequence = p.integer("sequence"),
          let modelId = p.string("model_id"),
          let artifactVersion = p.string("artifact_version"),
          let path = p.string("path"),
          let bytes = p.integer("bytes"),
          let sha256 = p.string("sha256"),
          let upstreamRepo = p.string("upstream_repo"),
          let upstreamRevision = p.string("upstream_revision"),
          let upstreamFilename = p.string("upstream_filename"),
          let architecture = p.string("architecture"),
          let quantization = p.string("quantization"),
          let runtimeBuildIds = p.stringArray("runtime_build_ids"),
          let minAppBuild = p.integer("min_app_build"),
          let maxAppBuild = p.integer("max_app_build"),
          let promptVersion = p.string("prompt_version"),
          let licenseNoticeId = p.string("license_notice_id"),
          let issuedAt = p.string("issued_at"),
          let expiresAt = p.string("expires_at") else {
      return reject(invalid, "missing or mistyped field")
    }
    guard schema == 1 else { return reject(invalid, "unknown schema") }
    guard sequence >= 1 else { return reject(invalid, "sequence") }
    guard modelId == NamuConstants.modelId else { return reject(invalid, "model_id") }
    guard isArtifactVersion(artifactVersion) else { return reject(invalid, "artifact_version") }
    guard isValidPath(path) else { return reject(invalid, "path") }
    guard bytes >= 1, bytes <= 8_000_000_000 else { return reject(invalid, "bytes") }
    guard NamuPattern.isSha256Hex(sha256) else { return reject(invalid, "sha256") }
    guard !upstreamRepo.isEmpty, !upstreamFilename.isEmpty,
          NamuPattern.isLowerHex(upstreamRevision, length: 40) else {
      return reject(invalid, "upstream fields")
    }
    guard let issued = parseTimestamp(issuedAt), let expires = parseTimestamp(expiresAt), expires > issued else {
      return reject(invalid, "timestamps")
    }

    let payloadSha256 = Data(SHA256.hash(data: payload)).namuHex

    // 6 + 7. Sequence and validity apply to remote metadata only (SIG-004).
    if ctx.source == .update {
      if expires - issued > maxValidityMs { return reject(invalid, "validity too long") }
      if issued > ctx.nowMs + clockSkewMs || ctx.nowMs >= expires {
        return reject(invalid, "expired or not yet valid")
      }
      if sequence < ctx.highestSequence { return reject(invalid, "sequence replay") }
      if sequence == ctx.highestSequence, let known = ctx.highestSequencePayloadSha256, !known.isEmpty,
         known != payloadSha256 {
        return reject(invalid, "same sequence, different payload")
      }
    }

    // Compatibility failures are MODEL_INCOMPATIBLE.
    let incompatible = NamuError.modelIncompatible
    guard ctx.profile.architectures.contains(architecture),
          ctx.profile.quantizations.contains(quantization) else {
      return reject(incompatible, "architecture/quantization")
    }
    guard runtimeBuildIds.contains(ctx.runtimeBuildId) else { return reject(incompatible, "runtime build") }
    guard minAppBuild <= maxAppBuild, ctx.appBuild >= minAppBuild, ctx.appBuild <= maxAppBuild else {
      return reject(incompatible, "app build range")
    }
    guard ctx.promptVersions.contains(promptVersion) else { return reject(incompatible, "prompt version") }
    guard ctx.licenseNoticeIds.contains(licenseNoticeId) else { return reject(incompatible, "license notice") }

    // 8. Known-bad artifacts are never offered again.
    if ctx.knownBad.contains(sha256) { return reject(NamuError.fileDamaged, "known-bad artifact") }

    let descriptor = ReleaseDescriptor(
      schema: schema, sequence: sequence, modelId: modelId, artifactVersion: artifactVersion,
      path: path, bytes: bytes, sha256: sha256, upstreamRepo: upstreamRepo,
      upstreamRevision: upstreamRevision, upstreamFilename: upstreamFilename,
      architecture: architecture, quantization: quantization, runtimeBuildIds: runtimeBuildIds,
      minAppBuild: minAppBuild, maxAppBuild: maxAppBuild, promptVersion: promptVersion,
      licenseNoticeId: licenseNoticeId, issuedAtMs: issued, expiresAtMs: expires)
    return .accepted(descriptor, payloadSha256: payloadSha256)
  }

  // MARK: - Field rules

  /// Standard base64 with padding: `^[A-Za-z0-9+/]*={0,2}$`, length % 4 == 0.
  static func decodeBase64(_ value: String?) -> Data? {
    guard let value else { return nil }
    let bytes = Array(value.utf8)
    guard bytes.count % 4 == 0 else { return nil }
    var padding = 0
    for (offset, byte) in bytes.enumerated() {
      if byte == UInt8(ascii: "=") {
        padding += 1
        guard offset >= bytes.count - 2 else { return nil }
      } else {
        guard padding == 0 else { return nil }
        let isAlphabet = (byte >= 0x41 && byte <= 0x5a) || (byte >= 0x61 && byte <= 0x7a)
          || (byte >= 0x30 && byte <= 0x39) || byte == 0x2b || byte == 0x2f
        guard isAlphabet else { return nil }
      }
    }
    return Data(base64Encoded: value)
  }

  private static func isPathCharacter(_ byte: UInt8) -> Bool {
    (byte >= 0x41 && byte <= 0x5a) || (byte >= 0x61 && byte <= 0x7a) || (byte >= 0x30 && byte <= 0x39)
      || byte == 0x2e || byte == 0x5f || byte == 0x2f || byte == 0x2d
  }

  /// `^[A-Za-z0-9._-]{1,64}$`
  static func isArtifactVersion(_ value: String) -> Bool {
    let bytes = Array(value.utf8)
    return (1...64).contains(bytes.count) && bytes.allSatisfy { isPathCharacter($0) && $0 != 0x2f }
  }

  /// `^[A-Za-z0-9._/-]{1,512}$`, relative, no empty, `.` or `..` segment.
  static func isValidPath(_ value: String) -> Bool {
    let bytes = Array(value.utf8)
    guard (1...512).contains(bytes.count), bytes.allSatisfy(isPathCharacter) else { return false }
    guard bytes.first != 0x2f else { return false }
    let segments = value.split(separator: "/", omittingEmptySubsequences: false)
    return segments.allSatisfy { !$0.isEmpty && $0 != "." && $0 != ".." }
  }

  /// RFC 3339 UTC, exactly `YYYY-MM-DDTHH:MM:SSZ`. Returns epoch milliseconds.
  static func parseTimestamp(_ value: String) -> Int64? {
    let b = Array(value.utf8)
    guard b.count == 20, b[4] == 0x2d, b[7] == 0x2d, b[10] == UInt8(ascii: "T"),
          b[13] == 0x3a, b[16] == 0x3a, b[19] == UInt8(ascii: "Z") else { return nil }
    func number(_ range: Range<Int>) -> Int64? {
      var out: Int64 = 0
      for i in range {
        guard b[i] >= 0x30, b[i] <= 0x39 else { return nil }
        out = out * 10 + Int64(b[i] - 0x30)
      }
      return out
    }
    guard let year = number(0..<4), let month = number(5..<7), let day = number(8..<10),
          let hour = number(11..<13), let minute = number(14..<16), let second = number(17..<19) else {
      return nil
    }
    guard (1...12).contains(month), minute <= 59, second <= 59 else { return nil }
    // ECMAScript accepts 24:00:00 as the end of the day; mirror the reference.
    guard hour <= 23 || (hour == 24 && minute == 0 && second == 0) else { return nil }
    let leap = (year % 4 == 0 && year % 100 != 0) || year % 400 == 0
    let daysInMonth: [Int64] = [31, leap ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    guard day >= 1, day <= daysInMonth[Int(month - 1)] else { return nil }
    // Days from civil (proleptic Gregorian), Howard Hinnant's algorithm.
    let y = month <= 2 ? year - 1 : year
    let era = (y >= 0 ? y : y - 399) / 400
    let yoe = y - era * 400
    let mp = (month + 9) % 12
    let doy = (153 * mp + 2) / 5 + day - 1
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy
    let days = era * 146_097 + doe - 719_468
    return ((days * 24 + hour) * 60 + minute) * 60_000 + second * 1000
  }

  static func formatTimestamp(ms: Int64) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime]
    formatter.timeZone = TimeZone(secondsFromGMT: 0)
    return formatter.string(from: Date(timeIntervalSince1970: Double(ms) / 1000))
  }
}

File: android/namu-core/src/main/kotlin/org/namuai/offline/core/descriptor/DescriptorVerifier.kt
package org.namuai.offline.core.descriptor

import org.namuai.offline.core.ErrorCodes
import org.namuai.offline.core.json.JsonValue
import org.namuai.offline.core.json.StrictJson
import org.namuai.offline.core.json.StrictJsonException
import org.namuai.offline.core.json.integer
import org.namuai.offline.core.json.string
import org.namuai.offline.core.json.stringArray
import org.namuai.offline.core.util.Hex
import java.security.MessageDigest
import java.time.DateTimeException
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.Base64

enum class DescriptorSource(val wire: String) {
    BUNDLED("bundled"),
    UPDATE("update");

    companion object {
        fun fromWire(value: String): DescriptorSource? = values().firstOrNull { it.wire == value }
    }
}

data class ReleaseKey(val keyId: String, val publicKeyB64: String)

/** Accepted architecture/quantization pairs (contract §4, REL-001 fixture profile). */
data class CompatibilityProfile(val architectures: Set<String>, val quantizations: Set<String>) {
    companion object {
        val PRODUCTION = CompatibilityProfile(setOf("cohere2"), setOf("Q4_K_M"))

        /** `.internal` builds only. */
        val INTERNAL = CompatibilityProfile(
            setOf("cohere2", "llama"),
            setOf("Q4_K_M", "Q8_0", "F16", "F32"),
        )
    }
}

data class Descriptor(
    val schema: Long,
    val sequence: Long,
    val modelId: String,
    val artifactVersion: String,
    val path: String,
    val bytes: Long,
    val sha256: String,
    val upstreamRepo: String,
    val upstreamRevision: String,
    val upstreamFilename: String,
    val architecture: String,
    val quantization: String,
    val runtimeBuildIds: List<String>,
    val minAppBuild: Long,
    val maxAppBuild: Long,
    val promptVersion: String,
    val licenseNoticeId: String,
    val issuedAtMs: Long,
    val expiresAtMs: Long,
)

class VerifyContext(
    val keys: List<ReleaseKey>,
    val source: DescriptorSource,
    val appBuild: Long,
    val runtimeBuildId: String,
    val promptVersions: Set<String>,
    val licenseNoticeIds: Set<String>,
    val nowMs: Long,
    val highestSequence: Long,
    val highestSequencePayloadSha256: String?,
    val knownBad: Set<String>,
    val profile: CompatibilityProfile,
)

sealed class VerifyResult {
    class Accepted(val descriptor: Descriptor, val payloadSha256: String) : VerifyResult()

    /** [reason] is for diagnostics/tests only; it never contains payload text. */
    class Rejected(val code: String, val reason: String) : VerifyResult()
}

/**
 * Signed release descriptor verification (SIG-001…005), step for step the
 * same order as model-release/descriptor/verify.mjs so both agree on every
 * shared vector (T08, T29, T30). The payload is parsed only AFTER the
 * signature over its exact bytes has been verified (SIG-002).
 */
class DescriptorVerifier(private val ed25519: Ed25519Verifier) {

    fun verify(envelopeBytes: ByteArray, ctx: VerifyContext): VerifyResult {
        if (envelopeBytes.isEmpty() || envelopeBytes.size > MAX_ENVELOPE_BYTES) {
            return invalid("envelope size")
        }
        val envelope = try {
            StrictJson.parseUtf8(envelopeBytes)
        } catch (e: StrictJsonException) {
            return invalid("envelope parse")
        }
        if (envelope !is JsonValue.Obj) return invalid("envelope not an object")
        val keyId = envelope.string("key_id")
        val payload = envelope.string("payload_b64")?.let { decodeBase64(it) }
        val signature = envelope.string("signature_b64")?.let { decodeBase64(it) }
        if (keyId.isNullOrEmpty() || payload == null || signature == null) {
            return invalid("envelope fields")
        }
        if (payload.isEmpty() || payload.size > MAX_PAYLOAD_BYTES) return invalid("payload size")
        if (signature.size != 64) return invalid("signature size")
        val key = ctx.keys.firstOrNull { it.keyId == keyId } ?: return invalid("unknown key")
        val publicRaw = decodeBase64(key.publicKeyB64)
        if (publicRaw == null || publicRaw.size != 32) return invalid("bundled key")
        if (!ed25519.verify(publicRaw, payload, signature)) return invalid("signature")

        // Only verified bytes are parsed from here on (SIG-002).
        val p = try {
            StrictJson.parseUtf8(payload)
        } catch (e: StrictJsonException) {
            return invalid("payload parse")
        }
        if (p !is JsonValue.Obj) return invalid("payload not an object")

        val schema = p.integer("schema") ?: return missing("schema")
        val sequence = p.integer("sequence") ?: return missing("sequence")
        val modelId = p.string("model_id") ?: return missing("model_id")
        val artifactVersion = p.string("artifact_version") ?: return missing("artifact_version")
        val path = p.string("path") ?: return missing("path")
        val bytes = p.integer("bytes") ?: return missing("bytes")
        val sha256 = p.string("sha256") ?: return missing("sha256")
        val upstreamRepo = p.string("upstream_repo") ?: return missing("upstream_repo")
        val upstreamRevision = p.string("upstream_revision") ?: return missing("upstream_revision")
        val upstreamFilename = p.string("upstream_filename") ?: return missing("upstream_filename")
        val architecture = p.string("architecture") ?: return missing("architecture")
        val quantization = p.string("quantization") ?: return missing("quantization")
        val runtimeBuildIds = p.stringArray("runtime_build_ids") ?: return missing("runtime_build_ids")
        val minAppBuild = p.integer("min_app_build") ?: return missing("min_app_build")
        val maxAppBuild = p.integer("max_app_build") ?: return missing("max_app_build")
        val promptVersion = p.string("prompt_version") ?: return missing("prompt_version")
        val licenseNoticeId = p.string("license_notice_id") ?: return missing("license_notice_id")
        val issuedAt = p.string("issued_at") ?: return missing("issued_at")
        val expiresAt = p.string("expires_at") ?: return missing("expires_at")

        if (schema != 1L) return invalid("unknown schema")
        if (sequence < 1) return invalid("sequence")
        if (modelId != MODEL_ID) return invalid("model_id")
        if (!ARTIFACT_VERSION.matches(artifactVersion)) return invalid("artifact_version")
        if (!isValidPath(path)) return invalid("path")
        if (bytes < 1 || bytes > MAX_ARTIFACT_BYTES) return invalid("bytes")
        if (!SHA256_HEX.matches(sha256)) return invalid("sha256")
        if (upstreamRepo.isEmpty() || upstreamFilename.isEmpty() || !REVISION.matches(upstreamRevision)) {
            return invalid("upstream fields")
        }
        val issued = parseTimestamp(issuedAt)
        val expires = parseTimestamp(expiresAt)
        if (issued == null || expires == null || expires <= issued) return invalid("timestamps")

        val payloadSha256 = Hex.encode(MessageDigest.getInstance("SHA-256").digest(payload))

        if (ctx.source == DescriptorSource.UPDATE) {
            if (expires - issued > MAX_VALIDITY_MS) return invalid("validity too long")
            if (issued > ctx.nowMs + CLOCK_SKEW_MS || ctx.nowMs >= expires) {
                return invalid("expired or not yet valid")
            }
            if (sequence < ctx.highestSequence) return invalid("sequence replay")
            val pinned = ctx.highestSequencePayloadSha256
            if (sequence == ctx.highestSequence && !pinned.isNullOrEmpty() && pinned != payloadSha256) {
                return invalid("same sequence, different payload")
            }
        }

        if (architecture !in ctx.profile.architectures || quantization !in ctx.profile.quantizations) {
            return incompatible("architecture/quantization")
        }
        if (ctx.runtimeBuildId !in runtimeBuildIds) return incompatible("runtime build")
        if (minAppBuild > maxAppBuild || ctx.appBuild < minAppBuild || ctx.appBuild > maxAppBuild) {
            return incompatible("app build range")
        }
        if (promptVersion !in ctx.promptVersions) return incompatible("prompt version")
        if (licenseNoticeId !in ctx.licenseNoticeIds) return incompatible("license notice")
        if (sha256 in ctx.knownBad) return VerifyResult.Rejected(ErrorCodes.FILE_DAMAGED, "known-bad artifact")

        return VerifyResult.Accepted(
            Descriptor(
                schema, sequence, modelId, artifactVersion, path, bytes, sha256,
                upstreamRepo, upstreamRevision, upstreamFilename, architecture, quantization,
                runtimeBuildIds, minAppBuild, maxAppBuild, promptVersion, licenseNoticeId,
                issued, expires,
            ),
            payloadSha256,
        )
    }

    private fun invalid(reason: String) = VerifyResult.Rejected(ErrorCodes.SIGNATURE_INVALID, reason)
    private fun missing(field: String) = invalid("missing or mistyped $field")
    private fun incompatible(reason: String) = VerifyResult.Rejected(ErrorCodes.MODEL_INCOMPATIBLE, reason)

    companion object {
        const val MAX_ENVELOPE_BYTES = 65_536
        const val MAX_PAYLOAD_BYTES = 32_768
        const val MAX_ARTIFACT_BYTES = 8_000_000_000L
        const val MAX_VALIDITY_MS = 180L * 24 * 3600 * 1000
        const val CLOCK_SKEW_MS = 24L * 3600 * 1000
        const val MODEL_ID = "namu-aya-global"

        private val B64 = Regex("[A-Za-z0-9+/]*={0,2}")
        private val ARTIFACT_VERSION = Regex("[A-Za-z0-9._-]{1,64}")
        private val PATH = Regex("[A-Za-z0-9._/-]{1,512}")
        val SHA256_HEX = Regex("[0-9a-f]{64}")
        private val REVISION = Regex("[0-9a-f]{40}")
        private val TIMESTAMP = Regex("([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{2}):([0-9]{2}):([0-9]{2})Z")

        /** Standard base64 with padding; null on any deviation. */
        fun decodeBase64(value: String): ByteArray? {
            if (value.length % 4 != 0 || !B64.matches(value)) return null
            return try {
                Base64.getDecoder().decode(value)
            } catch (e: IllegalArgumentException) {
                null
            }
        }

        fun isValidPath(path: String): Boolean {
            if (!PATH.matches(path)) return false
            if (path.startsWith("/") || path.contains("//")) return false
            return path.split("/").all { it.isNotEmpty() && it != "." && it != ".." }
        }

        /** RFC 3339 UTC, exactly `YYYY-MM-DDTHH:MM:SSZ`; null when malformed. */
        fun parseTimestamp(value: String): Long? {
            val m = TIMESTAMP.matchEntire(value) ?: return null
            val g = m.groupValues
            return try {
                LocalDateTime.of(
                    g[1].toInt(), g[2].toInt(), g[3].toInt(),
                    g[4].toInt(), g[5].toInt(), g[6].toInt(),
                ).toEpochSecond(ZoneOffset.UTC) * 1000
            } catch (e: DateTimeException) {
                null
            }
        }

        /**
         * Reads fields from an envelope that THIS app already verified and stored
         * in its own journal. Never call this on bytes from the network.
         */
        fun payloadOfStoredEnvelope(envelopeBytes: ByteArray): JsonValue.Obj? {
            return try {
                val env = StrictJson.parseUtf8(envelopeBytes) as? JsonValue.Obj ?: return null
                val payload = env.string("payload_b64")?.let { decodeBase64(it) } ?: return null
                StrictJson.parseUtf8(payload) as? JsonValue.Obj
            } catch (e: StrictJsonException) {
                null
            }
        }
    }
}
